Port the core/code block to be used by the ReactNative mobile GB app

hypest opened the pull request:
## Description
This PR introduces the minimal set of changes to make the core/code block ready to be consumed by a ReactNative app, in the road to implement a "native" mobile GB editing experience as per [the Make post](https://make.wordpress.org/mobile/2018/03/21/gutenberg-on-mobile/).

Among the goals is to eventually be able to introduce patterns for how to share code between GB the webapp and GB the native mobile app. Please, feel free to open PRs against this one with proposals towards that goal.

## TODO before lifting the "In progress" label:
- [x] Make sure the (web) GB build and webapp work normally
- [x] Factorize the RN platform dependent files to minimize code duplication between Android and iOS
- [x] ~Have a proposal ready for how to change the `core/code` block definition so its high level definition is shared among the 3 platforms (web, Android, iOS)~
For the time being, the `.native.js` variants of key files (`index.js` among which is the most important ones) will serve as the method we provide compatibility with an RN app. The goal is still to come up with a "common language" down the road but we need to port more code before patterns on what the common language is emerge.

## Types of changes
* What types of changes does your code introduce?

Introduces `.native.js` variants of various key files. Those are picked up instead of the corresponding `.js` files when the code is imported in an RN build. The regular Gutenberg build just ignores the `.native.js` files.

The specialized `index.native.js` files play a key role: They are importing only the submodules needed by the RN implementation. These submodules will gradually be ported and only where appropriate.

* New feature (non-breaking change which adds functionality)

Ports the `core/code` block into a RN compatible component.

* Breaking change (fix or feature that would cause existing functionality to not work as expected)

None.

## Checklist:
- [x] My code is tested. (Yes, [tests are on the gutenberg-mobile repo](https://github.com/wordpress-mobile/gutenberg-mobile/pull/25))
- [x] My code follows the WordPress code style.
- [x] My code has proper inline documentation.

hypest committed:
Android specific files for the Code block

* CSS related imports are suppressed
* classnames are left in only to minimize diff among the files but do not play a role, nor style the elements

hypest committed:
Breaking change: Don't override the default JSX translation

So far, that's the only way I found to have the android-specific jest tests succeed.

hypest committed:
Add the platform specific files for iOS as well

hypest committed:
Set TextInput to multiline for Code block

hypest committed:
Hide the default TextInput underline on Android

hypest committed:
Expose registerBlockType, getBlockType to the RN app

hypest commented:
👋 @gziolo , please feel free to add more reviewers to this PR if you think it's good they keep an open eye to this PR. Thanks!

👋@maxme , feel free to add more `TODO` items to the list. Thanks!

aduth commented:
>Have a proposal ready for how to change the core/code block definition so it's high level definition is shared among the 3 platforms (web, Android, iOS)

May also tie into #2751 (registration of said high-level definition on the server).

gziolo commented:
> >Have a proposal ready for how to change the core/code block definition so it's high level definition is shared among the 3 platforms (web, Android, iOS)
> 
> May also tie into #2751 (registration of said high-level definition on the server).

In general, we are looking into having a way to define all blocks on the server (minus JS specific bits like `edit`, `save` or other function based properties). We will have to wrap this with REST API endpoint to make it easy to consume by the platforms that don't use HTML to expose data from the server.

hypest committed:
Use .native.js for common native code

hypest commented:
Collapsed (with 8ee1a4e) platform specific code into `.native.js` files instead of individual (android, ios) ones. Thanks @gziolo for the tip!

hypest committed:
Merge branch 'master' into rnmobile/port-code-block

hypest committed:
Correct configuration for the babel plugins settings

hypest committed:
transform-react-jsx for the production env as well

hypest committed:
No need for pragma override

Will define a `wp` global via RN and reuse the `createElement()` from `@gutenberg/component`.

hypest committed:
Ignore mobile native files when linting

hypest commented:
Update:

* Merged from master and fixed any conflict
* Lifted the #pragma override of the JSX Babel transformer plugin
* Added the mobile native source files (`.native.js` and friends) to the eslint ignore list. Those files are currently processed via Flow and Prettier on the RN project side.
* Tests now pass

hypest committed:
Revert "Ignore mobile native files when linting"

This reverts commit a3d5ab4af34941e5afb12eba3b620dce565a78e6.

hypest committed:
Comment out ununsed imports

hypest committed:
Fix codestyle

hypest committed:
Import is compatible anyway, uncomment it

hypest committed:
Remove the React import. The RN project auto-imports it.

hypest commented:
Update:

* Re-established eslint for `.native.js` files and fixed all error occurrences
* Removed the React import from the `.native.js` files and instead, auto-importing React via the RN project's build pipeline

hypest committed:
Limit native code to the edit function alone

hypest commented:
Update: added a commit (685d178) that factors out the _view_ related definition out of the core/code block, into platform specific modules.

Can't tell yet if that's going to be scalable or usable in the rest of the blocks but, would love to hear some initial comments/thoughts on this.

WDYT @gziolo ?

hypest committed:
Merge branch 'master' into rnmobile/port-code-block

hypest committed:
Expose getBlockContent to RN

hypest committed:
RN app tests need createElement exported

hypest committed:
Export serialize from serializer

hypest committed:
Load native version of plain-text's CSS module

hypest committed:
Make it clear that RN uses styles, not classnames

maxme committed:
Merge branch 'rnmobile/react-native-mobile-integration-first-steps' into rnmobile/port-code-block

maxme committed:
Use the i18n package from @wordpress

maxme commented:
Note: I changed the base branch to point to our future feature branch: `rnmobile/react-native-mobile-integration-first-steps`

hypest committed:
Remove some commented out code

hypest committed:
Utils are not currently used on native

hypest committed:
'Border' property not supported in shorthand form

hypest committed:
Use hypens in css class names as traditionally

hypest committed:
remove unused code

hypest committed:
No need for platform-specific factory.js

hypest committed:
Merge branch 'master' into rnmobile/port-code-block

hypest committed:
Move native files after update from master

hypest committed:
Codestyle fix

hypest committed:
GB isn't using Prettier so, no need for the pragma

hypest committed:
Comment on the need to import the `edit` function

hypest commented:
After some discussion with @gziolo I've :
* cleaned up the "todo-looking" comments
* minimized the PR even further
* changed the target branch back to `master`
* updated from `master` and fixed any incompatibilities (e.g. due to file structure change)
* updated the PR description
* lifted the `[Status] In Progress` label.

gziolo commented on core-blocks/code/edit.native.js:
```diff
@@ -0,0 +1,28 @@
+ */
+import { PlainText } from '../../blocks';
+
+export function edit( { attributes, setAttributes, style } ) {
```
It can be exported as `export default function ...`.

hypest replied:
Makes sense, done with 0edfb18.

gziolo commented on core-blocks/code/index.js:
```diff
@@ -4,13 +4,12 @@
  * Internal dependencies
  */
-import './editor.scss';
+import { edit } from './edit'; // import the specialized `edit` function. Is is different between the web and native mobile.
```
This comment can be removed as we should probably do the same for all `edit` implementations :)

`edit` can be exported as default to have:

```js
import edit from './edit';

// and later
{ 
     ...
    edit,
    ...
}
```

hypest replied:
The comment can be considered trivial if one reads the PR and has an understanding of the `.native.js` pattern. I had added it to help people pick up the pattern but, i'm perfectly fine with removing it 👍 . Done with 9b68ba3.

gziolo commented on core-blocks/code/edit.native.js:
```diff
@@ -0,0 +1,28 @@
+
+export function edit( { attributes, setAttributes, style } ) {
+	return (
+		<View>
```
Shouldn't be `style` applied to `View` rather than `PlainText` to mirror web implementation?

hypest replied:
Yes and no. Unfortunately, styling in RN is not cascading by default and we will need more work on establishing a pattern on how to pass and style the blocks/components/containers.

In this particular case for example, the style includes the `fontFamily` property which is supported by the `PlaintText` component (because it's a `TextInput` under the hood) but by the `View` component.

All in all, styling should be considered `alpha` at this stage as well and i'd suggest we work on it in a different PR, if that's OK with you too @gziolo . I'll update the PR description to include some relative info.

gziolo replied:
That would be awesome to have it documented in the code next to RN version of the `edit` component.

hypest replied:
Done with 0377c73.

gziolo commented on blocks/plain-text/style.native.scss:
```diff
@@ -0,0 +1,9 @@
+.blocks-plain-text {
```
Did you check if it isn't included in web build by accident?

hypest replied:
Not extensively, I didn't. What I only tested was modifying values in the style file and see if GB-web picks those up, visually. Tried out adding a `background-color` for example. Any practical way to test it out more thoroughly perhaps?

gziolo replied:
I can test it myself before merging, but I think if you put `background-color: pink` in the RN version, it shouldn't get applied in the web version. That's all we need to ensure :)

hypest replied:
That's grand, that was the test I did and it succeeded (color wasn't picked up) so, feel free to test is your side as well, thanks!

gziolo reviewed (commented):
I left some minor comments. In general, I'm okey with the proposed solution, because everything happens in RN specific files. I think it would be a good idea to move all `edit` components to their own files anyway. We can help with that if others agree that it is a way to move forward.

@mtias, @aduth, @youknowriad any thoughts?

hypest committed:
Export the edit function as default

hypest committed:
Remove reduntant line comment

hypest committed:
Import block using the new pattern

hypest commented:
Thanks for the review @gziolo , ready for another pass! FYI, I made a minor change in the meantime (5c93331).

hypest committed:
Add a comment about styling

gziolo reviewed (approved):
I tested web version and it still works properly. I would defer the final decision to @aduth and @youknowriad, but speaking myself it's good to go.

gziolo committed:
Use object property shorthand for `edit`

gziolo commented:
Let's get it and iterate with further PRs.

mtias commented:
Good to see this happening :)
